syntax = "proto3";

package amazon.ups;

// Common types
message Coordinate {
  int32 x = 1;
  int32 y = 2;
}

message ItemInfo {
  string item_name = 1;
  int32 quantity = 2;
}

// Envelope for Amazon → UPS messages
message AmazonToUPS {
  repeated int64 acks = 124;
  repeated RequestPickup request_pickup = 1;
  repeated Redirect redirect = 2;
  repeated Cancel cancel = 3;
  repeated LoadReady load_ready = 4;
}

// Envelope for UPS → Amazon messages
message UPSToAmazon {
  repeated int64 acks = 124;
  repeated PickupResp pickup_ack = 1;
  repeated RedirectResp redirect_ack = 2;
  repeated CancelResp cancel_ack = 3;
  repeated TruckArrived truck_arrived = 4;
  repeated DeliveryStarted delivery_started = 6;
  repeated DeliveryComplete delivery_complete = 7;
  repeated Returned returned_msg = 8;
}

// Amazon → UPS messages
message RequestPickup {
  int64 seqnum = 123;
  string ups_user_id = 1;
  repeated ItemInfo items = 2;
  int64 order_id = 3;
  int32 warehouse_id = 4;
  Coordinate user_destination = 5;
}

message Redirect {
  int64 seqnum = 123;
  int64 package_id = 1;
  Coordinate new_destination = 2;
}

message Cancel {
  int64 seqnum = 123;
  int64 package_id = 1;
}

message LoadReady {
  int64 seqnum = 123;
  int64 package_id = 1;
  int32 truck_id = 2;
}

message ReturnAck {
  int64 seqnum = 123;
  int64 package_id = 1;
}

// UPS → Amazon messages
message PickupResp {
  int64 seqnum = 123;
  int64 package_id = 1;
  int64 order_id = 2;
  int32 truck_id = 3;
}

message RedirectResp {
  int64 seqnum = 123;
  int64 package_id = 1;
  bool success = 2;
  string reason = 3;
}

message CancelResp {
  int64 seqnum = 123;
  int64 package_id = 1;
  bool success = 2;
  string reason = 3;
}

message TruckArrived {
  int64 seqnum = 123;
  int64 package_id = 1;
  int32 truck_id = 2;
  int32 warehouse_id = 3;
}

message DeliveryStarted {
  int64 seqnum = 123;
  int64 package_id = 1;
}

message DeliveryComplete {
  int64 seqnum = 123;
  int64 package_id = 1;
}

message Returned {
  int64 seqnum = 123;
  int64 package_id = 1;
  int32 warehouse_id = 3;
}

message ReturnDelivered {
  int64 seqnum = 123;
  int64 package_id = 1;
  Coordinate destination = 2;
}
